Renesas: Smartbond: Add power management for USB

kasjer opened the pull request:
Smartbond does not have dedicated USB clock.
For USB to work PLL needs to be turned on.
To allow for flexible configuration artificial USB clock
is added that can be operated via clock_control subsystem.
This new clock turns on PLL when USB subsystem is enabled.
PLL can also be request in DT if application requires
more speed.

PLL can be automatically turned off when USB enters suspend
state and application did not requested it.

Few fixes were also added:
- VBUS interrupt was not enabled due to typo
- EP0 reception was not enabled again if it was disabled for a while
- EP0 RX FIFO could be accidentally flushed
- Sending of USB_DC_EP_DATA_OUT is not delayed till all interrupts from USB are handled this fixes race condition when transfer was started in interrupt but some other previous conditions were not checked yet.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

kasjer committed:
drivers: clock_control: smartbond: Add USB clock

Smartbodn does not have dedicated USB clock.
For USB to work PLL needs to be turned on.
To allow for flexible configuration artificial USB clock
is added that can be operated via clock_control subsystem.
This new clock turns on PLL when USB subsystem is enabled.
PLL can also be request in DT if application requires
more speed.

PLL can be automatically turned off when USB enters suspend
state and application did not requested it.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

kasjer committed:
drivers: usb_device: smartbond: Add power management

Add support for automatic control of PLL for USB subsystem.

This utilized virtual USB clock present in clock control.
USB driver enables PLL when device is attached or USB
sends resume signaling.

PLL can be turned off (if application does not requested it)
when device is detached (application request)
or USB host sends suspend request.

When VBUS is not present or application did not started USB yet
PD_DOMAIN_SYS is also not acquired that allows for deep
sleep. When USB is active deep sleep will never be activated.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

ioannis-karachalios commented:
We should also use PM constraints to prevent sleep as long as USB is present. The latter is connected to `VBUS IRQ` that should trigger `usb_dc_smartbond_vbus_changed` to be called (there is already TODO marks).

ioannis-karachalios reviewed (changes requested):
The USB block resides in PD_SYS and so we should use PM constraints to prevent sleep as long as a USB device is attached/present. The latter is bound to VBUS IRQ that should trigger `usb_dc_smartbond_vbus_changed` (there are already TODO marks).

ioannis-karachalios commented on drivers/clock_control/clock_control_smartbond.c:
```diff
@@ -231,8 +244,15 @@ static inline int smartbond_clock_control_off(const struct device *dev,
 	case SMARTBOND_CLK_PLL96M:
-		da1469x_clock_sys_pll_disable();
+		pll_requests &= ~(1 << (clk - SMARTBOND_CLK_PLL96M));
+		if (pll_requests == 0) {
```
We should expected conflict with [that ](https://github.com/zephyrproject-rtos/zephyr/pull/73238/commits/b68d74e952be6505bbda59dfc5a6180d82ae2d68#diff-fc167ec9957914ba118ad236eb3862f8736edaf47c0c69a7422eaa4fbee28267R230) PR. I am fine with that approach but we should also call `da1469x_clock_check_device_div1_clock` before we close PLL and the clock is transparently switched back to XTAL32M. The mentioned HAL API also checks if USB is active and aborts. To avoid multiple paths I suggest that we open a HAL PR and remove code lines 143-151. That should be


```
    /* Exercise if USB is enabled and clocked by PLL */
    if (USB->USB_MCTRL_REG & USB_USB_MCTRL_REG_USBEN_Msk) {

        uint32_t clk_ctrl_reg = CRG_TOP->CLK_CTRL_REG;
        if (((clk_ctrl_reg & CRG_TOP_CLK_CTRL_REG_USB_CLK_SRC_Msk) == 0) &&
            ((clk_ctrl_reg & CRG_TOP_CLK_CTRL_REG_SYS_CLK_SEL_Msk) == 3)) {
                return true;
            }
    }
```

ioannis-karachalios replied:
What is more, I think the [on-off manager](https://docs.zephyrproject.org/latest/services/resource_management/index.html) would be a good fit here. If you think that it'd be a cheaper solution we could take advantage of it.

kasjer replied:
@ioannis-karachalios it looks like *on-off manager* is present in Zephyr for over 4 years and is not used much except for few place in NRF related code. I think, I tried to incorporate this API in clock control for DA1469x at first but I abandoned this approach hence inclusion of onoff.h in our clock control that is not really used.

ioannis-karachalios replied:
Understood. If so, let's continue with the current approach.

ioannis-karachalios replied:
You need to re-base and resolve conflicts in the clock manager. I' have created a HAL PR that removes the USB check from `da1469x_clock_check_device_div1_clock` as a virtual USB clock is now introduced.

kasjer replied:
thanks @ioannis-karachalios , I will rebase, currently I'm adding code that takes VBUS into consideration. I though that when there is no VBUS whole USB state machine can be stopped and additionally PD_SYS could be released so device can go to deep sleep if everything else is ready. When VBUS is present and host made suspend condition on USB lines, PLL can be stopped if not requested by non-USB code (current code in this PR already does it) but PD_SYS will be held. Third condition is when VBUS is present but application requested detach in this state I think driver can also release PD_SYS (but this I have to double check).

ioannis-karachalios replied:
As far as the third condition is concerned, that is when app. requests detach, It makes sense that the USB block should be disabled and PD_SYS should be released.

ioannis-karachalios replied:
As discussed before, `da1469x_clock_check_device_div1_clock ` has been updated and should no longer exercise the USB clock source.

kasjer committed:
drivers: usb_device: smartbond: Fix VBUS interrupt enable

Code was enabling USB_IRQn twice. One of the calls clearly
should be enabling VBUS_IRQn.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

kasjer committed:
drivers: usb_device: smartbond: Fix reenable of EP0

When EP0 was disabled EP0 was setup to ignore setup and
out packets.
In usb_dc_ep_enable() function mask was never cleared.
It resulted in USB stop functioning after application
disabled endpoint (usb_disable()) and then tried
to enable usb again.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

kasjer committed:
drivers: usb_device: smartbond: Fix unnecessary flush

Switching from write to read on EP0 is done when data is
transmitted.
Additionally receiver is enabled when NAK is created in
response to OUT token. This may happen when host sent
ZLP before everything was transmitted.
This case flushed receiver FIFO which could lead to
setup packet corruption.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

kasjer committed:
drivers: usb_device: smartbond: Delay message USB_DC_EP_DATA_OUT

Callback with USB_DC_EP_DATA_OUT was called directly from
interrupt handler (handle_ep0_rx()) while rest of the interrupt
conditions were not checked yet.
Some interfaces started next transfer directly from this callback
resulting in case where TX endpoint was still marked as busy
and transfer was eventually abandoned.

Now message is sent from all affected endpoints after every USB
pending interrupt is handled.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

kasjer committed:
drivers: usb_device: smartbond: Fix EP OUT after resume

When device was resumed from sleep OUT endpoints were most
likely not enabled due to condition that checked pending transfer.
Configured OUT endpoints should be enabled (ready for RX)
if they were enabled by interfaces.

Now separate bit field enabled is added and check to
see if OUT endpoint should be ready for reception after
sleep is done.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

andrzej-kaczmarek commented on drivers/usb/device/usb_dc_smartbond.c:
```diff
@@ -1277,6 +1279,7 @@ int usb_dc_ep_enable(const uint8_t ep)
 		REG_SET_BIT(USB_MAMSK_REG, USB_M_TX_EV);
 		ep_state->regs->epc_in |= USB_USB_EPC2_REG_USB_EP_EN_Msk;
 	}
+	ep_state->enabled = true;
```
= 1

kasjer replied:
fixed

ioannis-karachalios commented on drivers/usb/device/usb_dc_smartbond.c:
```diff
@@ -890,36 +914,78 @@ static void usb_dc_smartbond_isr(void)
-		 */
+	if (!dev_state.vbus_present && vbus_present) {
+		dev_state.vbus_present = true;
+		da1469x_pd_acquire(MCU_PD_DOMAIN_SYS);
```
We should better use PM constraints to totally prevent sleep. There is no benefit in initiating the PM process and end up entering the WFI state within `da1469x_enter_sleep` (because of the PD_SYS reference counter being > 0). Besides, when a USB device is attached, VBUS should become greater than VBAT and so the entire SoC will be powered by VBUS.

kasjer replied:
@ioannis-karachalios instead of acquiring PD_SYS code now uses pm_policy_state_lock_put/get to prevent suspend state when PM is in use.

ioannis-karachalios commented on drivers/usb/device/usb_dc_smartbond.c:
```diff
@@ -890,36 +914,78 @@ static void usb_dc_smartbond_isr(void)
+		usb_clock_off();
 		dev_state.status_cb(USB_DC_DISCONNECTED, NULL);
-		/* TODO: Add PD release when available */
+		da1469x_pd_release(MCU_PD_DOMAIN_SYS);
```
See previous comment on PM constraints usage.

ioannis-karachalios commented:
Not sure if this should be handled now or we should wait for this PR to be merged first. There is a necessity for the following items to be updated:
1. DMA properties should be retrieved using the `dmas` DT property. The relevant [fix ](https://github.com/zephyrproject-rtos/zephyr/pull/73822)is already merged upstream. That is, TX/RX channels and USB slot configurations should be retrieved at initialization phase ([for instance](https://github.com/zephyrproject-rtos/zephyr/pull/73747/files#diff-8a70974e4b68b47cb59e0ba6253847031c56e6b55dc21e273a9cebc72e887b66R1273)).
2. DMA channels should be acquired and released using `dma_request_channel` and `dma_release_channel` respectively where necessary.
3. Add validity check for the DMA channels selection ([for reference](https://github.com/zephyrproject-rtos/zephyr/pull/73747/files#diff-8a70974e4b68b47cb59e0ba6253847031c56e6b55dc21e273a9cebc72e887b66R820))
4. The DMA accelerator should not be used directly, instead the corresponding DMA driver API class should be used (dma_start, dma_stop, dma_config, dma_reload).

kasjer commented:
Latest force-push is just rebase withough changes with hope that twister works on main

ydamigos commented on drivers/usb/device/usb_dc_smartbond.c:
```diff
@@ -121,16 +124,20 @@ struct smartbond_ep_state {
 	uint8_t nfsr;
 	usb_dc_status_callback status_cb;
 	struct smartbond_ep_state ep_state[2][4];
+	/** Btimask of EP OUT endpoints that received data during interrupt */
```
```suggestion
	/** Bitmask of EP OUT endpoints that received data during interrupt */
```

kasjer replied:
Typo fixed

jfischer-no commented on drivers/usb/device/usb_dc_smartbond.c:
```diff
@@ -892,34 +917,80 @@ static void usb_dc_smartbond_isr(void)
+	if (attached && vbus_present) {
+		dev_state.attached = true;
+		dev_state.vbus_present = true;
+#if defined(CONFIG_PM)
```
It looks like pm_policy_state_lock_get() can always build, is `#if defined(CONFIG_PM)` required here and below?

kasjer replied:
Thanks for advice, suggestion applied

jfischer-no commented on drivers/usb/device/usb_dc_smartbond.c:
```diff
@@ -914,6 +914,16 @@ static void usb_dc_smartbond_isr(void)
 	}
+
+	for (int i = 0; dev_state.ep_out_data && i < 4; ++i) {
+		uint8_t mask = 1 << i;
```
`const uint8_t mask = BIT(i);` ?

kasjer replied:
Fixed, thanks

ioannis-karachalios commented:
@nordic-krch is there any feedback from your side? If not, can you please approve? Thanks.
